README.md
#  Gtypist + record

## WHAT IS?

gnu-typist + 練習記録パッチ。

## 動作

練習ステージをひとつ終了したらその記録を ~/.gtypist に吐き出すよう、
gtypist を変更。

gtypist\_swing.rb あるいは gtypist\_tk.rb は定期的に ~/.gtypist を見て、
その更新状況をウィンドウ上に表示する。

gtypist.sh は記録を残すように改造した gtypist と共に viewer を立ち上げ、
gtypist の終了とともに viewer を終わらす。

## INSTALL on LINUX

情報センターPCへインストールの場合、

```sh
$ tar xfz gtypist-2.9.2.tar.gz
$ cd gtypist-2.9.2
$ ./configure --prefix=/edu
$ patch -p1 <../patch
$ make
$ make install
```

## INSTALL on OSX

```sh
$ tar xfz gtypist-2.9.2.tar.gz
$ cd gtypist-2.9.2
$ patch -p1 <../patch.osx
$ ./configure
$ patch -p1 <../patch
$ make
$ make install
```

## TODO

* (bug)swing ウィンドウが開いてもフォーカスを奪わないように。

* (bug)gtypist をインタラプトで終了すると、swing ウィンドウを取り残す。

* 2013-04-02, インストールの方法。

* gtypist upload 機能。

* gtypist をスタート時に自動起動し、gtypist を終了したらビューアも消えるように。

* gtypist_tk.rb から記録をアップロードする。

* gtypist_tk.rb のムダループはなくせるか？

* osx でビルド出来ない。
  -e Error:  both library and header files for the ncursesw library
  are required to build this package.  See INSTALL file for further
  information. On Debian/Ubuntu you need to install libncursesw5-dev.

 => (solved) Homebrew のパッチを入れる。

---
programmed by Hiroshi Kimura, 2013-03-22.


ChangeLog
2013-03-18  Hiroshi Kimura  <hiroshi.kimura.0331@gmail.com>

	* src/gtypist.c: /home/t/hkimura/以下ではなく、${HOME}/.gtypist に
	記録を残す。

	* gtypist_tk.rb:


gtypist_swing.rb
#!/usr/bin/env jruby
# -*- mode: ruby; coding: utf-8 -*-
# programmed by Hiroshi Kimura, 2012-04-22.
#
# VERSION:
#
# 本日の gtypist のスコアを表示する学生向けスクリプト。
#
# TODO: 2012-04-25, クリアしたステージを刻々と表示するように。
#  => done 2012-05-03.
#
# isc で実行すると非常に遅い、か? 再チェック。
#
DEBUG=false
INTERVAL=5

include Java
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTextArea

def debug(s)
  puts "debug: "+s if DEBUG
end

M={"Jan"=>"01", "Feb"=>"02", "Mar"=>"03", "Apr"=>"04",
   "May"=>"05", "Jun"=>"06", "Jul"=>"07", "Aug"=>"08",
   "Sep"=>"09", "Oct"=>"10","Nov"=>"11","Dec"=>"12"}

def D(d)
  if d.to_i<10
    "0"+d
  else
    d
  end
end

class MyApp
  def initialize
    @displayed=Hash.new
    frame=JFrame.new("gtypist")
    frame.set_default_close_operation(JFrame::EXIT_ON_CLOSE)
    panel=JPanel.new

    @text_area=JTextArea.new(20,25)
    @text_area.set_editable(false)
    sc=JScrollPane.new
    sc.setViewportView(@text_area)
    panel.add(sc)
    frame.add(panel)
    frame.pack
    frame.set_visible(true)
  end

  # FIXME: icome にアップロード。
  def update(records)
    records.each do |r|
      debug "record:#{r}"
      if @displayed.member?(r)
        next
      else
        @text_area.append(r+"\n")
        @displayed[r]=true
      end
    end
  end
end

def cleared_stages(fname)
  ret=Array.new
  last_stage=""
  File.foreach(fname) do |line|
    line=line.chomp
    wday,month,date,time,year,rest=line.split(/\s+/,6)
    if rest=~/with (\d+)\.(\d+)%/
      if "#{$1}.#{$2}".to_f <= 3.0
        ret.push last_stage
      end
    else
      last_stage=line
    end
  end
  ret
end

# main starts here.
begin
  app=MyApp.new
  fname = File.join(ENV['HOME'],'.gtypist')
  while not File.exists?(fname)
    sleep(INTERVAL)
  end
  while true
    stages = cleared_stages(fname)
    app.update(stages)
    sleep(INTERVAL)
  end

rescue =>e
  puts e.message
end


gtypist_today.rb
#!/usr/bin/env jruby
# -*- mode: ruby; coding: utf-8 -*-
# programmed by Hiroshi Kimura, 2012-04-22.
#
# 本日の gtypist のスコアを表示する学生向けスクリプト。
#
# TODO: 2012-04-25, クリアしたステージを刻々と表示するように。
# 	=> done 2012-05-03.
#
# this is a part of icome4/utils
# VERSION: 0.6.6
#
# isc で実行すると非常に遅い。
# tk で書き直し。
#
# JRuby can not.
#DEBUG=(RUBY_PLATFORM=~/darwin/)
DEBUG=File.directory?("/Applications")
INTERVAL=5

include Java
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTextArea

def debug(s)
  puts "debug: "+s if DEBUG
end

M={"Jan"=>"01", "Feb"=>"02", "Mar"=>"03", "Apr"=>"04",
   "May"=>"05", "Jun"=>"06", "Jul"=>"07", "Aug"=>"08",
   "Sep"=>"09", "Oct"=>"10","Nov"=>"11","Dec"=>"12"}

def D(d)
  if d.to_i<10
    "0"+d
  else
    d
  end
end

class MyApp
  def initialize
    frame=JFrame.new("gtypist today")
    frame.set_default_close_operation(JFrame::EXIT_ON_CLOSE)
    panel=JPanel.new

    @text_area=JTextArea.new(20,20)
    @text_area.set_editable(false)
    sc=JScrollPane.new
    sc.setViewportView(@text_area)
    panel.add(sc)
    frame.add(panel)
    frame.pack
    frame.set_visible(true)
  end

  def update(data)
    @text_area.set_text(data.reverse.join("\n"))
  end
end
#
# main starts here.
#
begin
  if DEBUG
    fname=File.join(ENV['HOME'],'.gtypist')
  else # isc
    fname=File.join("/home/t/hkimura/literacy/gtypist",ENV['USER'])
  end
# 起動直後にはまだファイルがないときもある。
#  raise "記録が見つかりません。#{fname}" unless File.exists?(fname)

  now=Time.now
  app=MyApp.new
  results=Array.new

  while true
    flag=cleared=rate=nil
    sleep(INTERVAL)
    next unless File.exists?(fname)
    IO.readlines(fname).reverse.each do |line|
      #    debug line
      wday,month,date,time,year,rest=line.chomp.split(/\s+/,6)
      unless now.day.to_i == date.to_i
        debug "not match: #{line}"
        next
      end
      if flag
        #      debug "found, rest:#{rest}"
        next unless rest=~/\*:(.*)$/
        #      debug "found2"
        score=$1
        results.push "#{cleared} #{score}" if score=~/_.+_.+_/
        flag=cleared=rate=nil
      else
        if rest=~/with (\d)\.(\d)% errors/
          rate="#{$1}.#{$2}".to_f
          if rate < 3.0
            cleared="#{year}-#{M[month]}-#{D(date)} #{time}"
            flag=true
          end
        end
      end
    end
    app.update(results)
  end

rescue =>e
  puts e.message
end


gtypist_tk.rb
#!/usr/bin/env ruby
# -*- mode: ruby; coding: utf-8 -*-
# programmed by Hiroshi Kimura, 2012-04-22.
#
# 本日の gtypist のスコアを表示する学生向けスクリプト。
# jruby で遅すぎるので、tk で書き直し。
#
# TODO: 2012-04-25, クリアしたステージを刻々と表示するように。
# 	=> done 2012-05-03.
#
# VERSION: 0.6.6
#
DEBUG=false
INTERVAL=3
require 'tk'

M={"Jan"=>"01", "Feb"=>"02", "Mar"=>"03", "Apr"=>"04",
   "May"=>"05", "Jun"=>"06", "Jul"=>"07", "Aug"=>"08",
   "Sep"=>"09", "Oct"=>"10","Nov"=>"11","Dec"=>"12"}

def debug(s)
  puts "debug: "+s if DEBUG
end

def osx?
  RUBY_PLATFORM=~/darwin/
end

def D(d)
  if d.to_i<10
    "0"+d
  else
    d
  end
end

class MyApp
  def initialize
    root=TkRoot.new
    root.title='gtypist'
    frame=TkFrame.new(root)
    @text=TkText.new(frame)
    @text.configure(width: 30,height: 10,background: 'white')
    @text.tag_configure('new',:background=>'red',:foreground=>'white')
    @text.pack(side: 'right')
    scr=TkScrollbar.new(frame)
    scr.pack(side: 'left',fill: 'y')
    @text.yscrollbar(scr)
    @text.configure(state:'disabled')
    frame.pack
  end

  def update(data)
    @text.configure(state:'normal')
    @text.delete('1.0','end')
    @text.insert('end',data.reverse.join("\n"))
    @text.tag_add('new',"#{data.length}.0",'end')
    @text.see('end')
    @text.configure(state:'disabled')
  end
end
#
# main starts here.
#
begin
  fname=File.join(ENV['HOME'],".gtypist")
  debug "fname:#{fname}"
  # 起動直後にはまだファイルがないときもある。
  #  raise "記録が見つかりません。#{fname}" unless File.exists?(fname)

  now=Time.now
  results=Array.new
  app=MyApp.new
  app.update(["programmed", "by", "hiroshi kimura."].reverse)
  lines=Array.new
  old_lines=Array.new

  Thread.new do
    while true
      flag=cleared=rate=nil
      results.clear
      sleep(INTERVAL)
      next unless File.exists?(fname)
      lines=IO.readlines(fname)
      next if lines==old_lines
      lines.reverse.each do |line|
        wday,month,date,time,year,rest=line.chomp.split(/\s+/,6)
        unless now.day.to_i == date.to_i
          debug "not today: #{line}"
          next
        end
        if flag
          next unless rest=~/\*:(.*)$/
          score=$1
          results.push "#{cleared} #{score}" if score=~/_.+_.+_/
          flag=cleared=rate=nil
        else
          if rest=~/with (\d)\.(\d)% errors/
            rate="#{$1}.#{$2}".to_f
            if rate < 3.0
              cleared="#{year}-#{M[month]}-#{D(date)} #{time}"
              flag=true
            end
          end
        end
      end
      app.update(results)
      old_lines=lines
    end
  end
  Tk.mainloop

rescue =>e
  puts e.message
end
